[FIX] base: allow user to modify its own preferences

awa-odoo opened the pull request:
The res.users model functions a bit uniquely since it should allow a user to modify its own parameters but turns off model edition priviledges by default. See "SELF_WRITEABLE_FIELDS".

This implies that ir_ui_view#_postprocess_access_rights method will by default turn off record edition, by automatically adding a 'edit="false"' attribute on the form node when the frontend calls 'get_views'.

This will in turn prevent the user from modifying its preferences as it will set the form view in readonly mode.
(This was apparently ignored pre-OWL, hence why we only have this issue now).

To fix this issue, we force the edition by manually setting 'edit="1"' on our "view_users_form_simple_modif" form.
Allowing the end-user to modify its own settings again.

Task-3067001

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/106125).

tde-banana-odoo commented:
@beledouxdenis seems legit isn't it ?

beledouxdenis commented:
> (This was apparently ignored pre-OWL, hence why we only have this issue now).

I just tested in 15.0 and the `edit="False"` was indeed already there, while the web client still allowed the users to edit the form fields.
![image](https://user-images.githubusercontent.com/5822488/203283353-ba268273-0411-4773-b1e5-0f827d237187.png)

> @beledouxdenis seems legit isn't it ?

There it does seems legit to me, though I would create a unit test, because of the combination of the tricky behavior of `res.users` and the fact having the users preferences not working and not covered in a test seems pretty critical to me. It's quite astonishing we see this only now.

tde-banana-odoo commented:
> Therefore it does seems legit to me, though I would create a unit test, because of the combination of the tricky behavior of `res.users` and the fact having the users preferences not working and not covered in a test seems pretty critical to me. It's quite astonishing we see this only now.

Does also seems legit :D .

awa-odoo commented:
Hello @beledouxdenis @tde-banana-odoo 

I added a unit test that makes sure a specified attribute value takes precedence over automatic ACL checks.
Although it's not exactly what I fixed here, so maybe that is not what you had in mind.

Thanks for the review & cheers!

beledouxdenis commented on odoo/addons/base/tests/test_views.py:
```diff
@@ -2213,6 +2213,47 @@ def test_modifier_attribute_using_context(self):
+        self.assertEqual(custom_form_node.get('create'), 'false')
+        self.assertEqual(custom_form_node.get('edit'), '1')
+        self.assertEqual(custom_form_node.get('delete'), 'false')
+
```
This unit test doesn't really test what your code change is fixing.

This unit test tests the fact `edit="1"` on the view node as the priority on the ACLs. But this was already working before your fix, this is not what your code change is fixing.

What should should test is that loading the User Preferences form view as the demo user allows the demo user to edit the fields. Given the fact `res.users` has a tricky behavior regarding the edition of its fields despite the ACLs normally prevent it.

So, to me, the unit test you should add should be placed in `odoo/addons/base/tests/test_res_users.py`, as it tests a behavior specific to `res.users` only.
In the unit test, you should load the view `view_users_form_simple_modif` through the `Form` test util, with the demo user as the demo user, then ensure the form is editable, for instance by checking changing fields, such as the user email or timezone, is allowed.

e.g. (untested code)
```py
demo_user = self.env.ref('base.user_demo')
with Form(demo_user.with_user(demo_user), view='base.view_users_form_simple_modif') as form:
    form.email = 'foo@bar.com'
self.assertEqual(demo_user.email, 'foo@bar.com')
```
And check that reverting your fix temporary, removing that `edit="1"` in that `view_users_form_simple_modif`, fails the above test.

awa-odoo replied:
Hello @beledouxdenis 

I tried a couple things today.

- The test that you provided.

Unfortunately, as the issue lies with the web framework that sets the form view as readonly, this test does not fail without my fix.

- A tour

After discussing this with TDE, we concluded that the best way to test it was a tour.
So I wrote a tour, in the mail module, as base does not have the tour assets.

Tested in local, worked great, life was good.

However, it's mandatory to set your tour tests as post_install.
(See https://github.com/odoo/odoo/blob/16.0/odoo/tests/common.py#L1822 ).
(We asked XDO, it has something to do with the database needing being fully initialized with all menus, otherwise it makes the tours impossible to debug, conclusion is we don't support it).

Making the tour post_install means the tour useless.
As "hr" module is installed, and it completely changes the flow of the user preferences (see my previous comment about that).

So I'm a bit lost here, I don't think I can write a test that will correctly cover this fix.

Unless you have an idea, maybe we could proceed without a test exceptionally?

Thanks again for the guidance.

awa-odoo replied:
Hello @beledouxdenis 

Actually after discussing this one more time with XDO, we came up with a different approach.
We keep the test in mail and adapt the steps in the hr module so that both flows can be tested.

At first I thought that it would mean we never really test the flow with only the mail module, but that's actually not true as single module builds are done during nightly builds on the runbot.

It also has the advantage of testing this flow when hr is installed as well, which seems like a good thing.

tde-banana-odoo commented:
@robodoo r+

Thanks for your work :)

std-odoo commented on addons/mail/tests/test_user_modify_own_profile.py:
```diff
@@ -0,0 +1,15 @@
+# -*- coding: utf-8 -*-
```
this test is very specific, can we move it to the `test_mail` module ? 
(to not bloat mail too much)

std-odoo reviewed (commented):
@awa-odoo Hi :) 

I have no comment, except that the test is super specific (so having 2 files (test + tour) in the mail module bloat it imo), can we move it to the `test_mail` module ?
